internal/olm/operator/registry/index_image.go

```diff
@@ -124,14 +151,306 @@ func (c IndexImageCatalogCreator) CreateCatalog(ctx context.Context, name string
 	return cs, nil
 }
 
+// getChannelHead retrieves the channel head from an array of entries
+func getChannelHead(entries []declarativeconfig.ChannelEntry) (string, error) {
+	nameMap := make(map[string]bool)
+	replacesMap := make(map[string]bool)
+
+	for i, _ := range entries {
+		nameMap[entries[i].Name] = true
+		if entries[i].Replaces != "" {
+			replacesMap[entries[i].Replaces] = true
+		}
+	}
+
+	// gets the CSV name that does not appear in any replaces field in the entries array
+	for key, _ := range nameMap {
+		if _, present := replacesMap[key]; !present {
+			return key, nil
+		}
+	}
+
+	// this should not be reached since there must be an edge to upgrade from
+	return "", errors.New("no channel head found")
+}
+
+// handleTraditionalUpgrade upgrades an operator that was installed using OLM. Subsequent upgrades will go through the setupFBCupdates function
+func handleTraditionalUpgrade(ctx context.Context, indexImage string, bundleImage string, channelName string) (string, error) {
+	// render the index image
+	render := action.Render{Refs: []string{indexImage}}
+	log.SetOutput(ioutil.Discard)
+	originalDeclCfg, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if err != nil {
+		return "", err
+	}
+
+	// render the bundle image
+	render = action.Render{Refs: []string{bundleImage}}
+	log.SetOutput(ioutil.Discard)
+	bundleDeclConfig, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if len(bundleDeclConfig.Bundles) != 1 {
+		return "", errors.New("bundle image must have at least one bundle")
+	}
+
+	if err != nil {
```

Comment:
Let's check for the error from `render.Run()` first and then ensure the correct bundle size. I suggest to move this check before `if len(bundleDeclConfig.Bundles) != 1`

```suggestion
```